SU2_CFD/src/iteration_structure.cpp

```diff
@@ -3023,27 +3023,16 @@ void CDiscAdjFEAIteration::SetRecording(COutput *output,
 
 void CDiscAdjFEAIteration::RegisterInput(CSolver *****solver_container, CGeometry ****geometry_container, CConfig **config_container, unsigned short iZone, unsigned short iInst, unsigned short kind_recording){
 
+  /*--- Register structural displacements as input ---*/
 
-  if (kind_recording == FEA_DISP_VARS){
+  solver_container[iZone][iInst][MESH_0][ADJFEA_SOL]->RegisterSolution(geometry_container[iZone][iInst][MESH_0], config_container[iZone]);
 
-    /*--- Register structural displacements as input ---*/
+  /*--- Register variables as input ---*/
 
-    solver_container[iZone][iInst][MESH_0][ADJFEA_SOL]->RegisterSolution(geometry_container[iZone][iInst][MESH_0], config_container[iZone]);
-
-    /*--- Register variables as input ---*/
-
-    solver_container[iZone][iInst][MESH_0][ADJFEA_SOL]->RegisterVariables(geometry_container[iZone][iInst][MESH_0], config_container[iZone]);
-
-  }
-
-  if (kind_recording == FEM_CROSS_TERM_GEOMETRY){
-
-    /*--- Register only structural displacements as input ---*/
-
-    solver_container[iZone][iInst][MESH_0][ADJFEA_SOL]->RegisterSolution(geometry_container[iZone][iInst][MESH_0], config_container[iZone]);
-
-  }
+  solver_container[iZone][iInst][MESH_0][ADJFEA_SOL]->RegisterVariables(geometry_container[iZone][iInst][MESH_0], config_container[iZone]);
 
+  /*--- Both need to be registered regardless of kind_recording for structural shape derivatives to work properly.
+        Otherwise, the code simply diverges as the FEM_CROSS_TERM_GEOMETRY breaks! (no idea why) for this term we register but do not extract! ---*/
```

Comment:
Do you have any new idea of why is this happening? It should not be the case, all the variables registered should be extracted. What happens if you add 
`solver_container[ZONE_STRUCT][INST_0][MESH_0][ADJFEA_SOL]->ExtractAdjoint_Variables()`
in 
`CDiscAdjFSIDriver::ExtractAdjoint()`
under
`(kind_recording == FEM_CROSS_TERM_GEOMETRY)`
does it still diverge then?

Reply:
Nope, no idea.
I do not know what would happen but if I extract those values I think they will be wrong, i.e. not the FEM_VARIABLE derivatives we want, due to the variables registered and iteration recorded at the time.

Reply:
I agree that they will probably be wrong, but maybe it's worth a try, given that the code diverges if that statements is not there.

Reply:
Without having read through the code, here's a guess what might be the reason.
The current AD implementation relies on one very important assumption: Derivatives/adjoints have to be extracted **in the very same order** as they have been registered.
So if you're obtaining wrong derivatives (that cause some divergence) and this can apparently be corrected by changing the registration, that's a big hint into that direction.

Just at a first glance: Could this reasoning apply here? If so, I'll have a more thorough look at it.

Reply:
Hi @oleburghardt,
I don't think so. These changes I'm proposing don't change the order of any extraction.
What's more the problem was in the cross terms and:
- These had one set of inputs and one set of outputs, so it would be hard to extract in the incorrect order;
- The spurious values appear in areas of the domain where we know everything has to be zero (those cross-terms only depend on boundary values) and when I tried setting those areas to zero manually I got correct derivatives, i.e. the values that are not spurious are correct and in the correct place.
That being said you are more than welcome to have a look.